### PR TITLE
Refactor periodic boundaries 🍩

### LINKED ISSUE #52: Add more natural choice of periodic boundaries
Currently, the "periodic" setting in SurfaceCode means periodic in all three directions. This is useful but unnatural. The setting should be changed to "all_periodic" and a new setting ("periodic") added with boundaries that are periodic in x and y but not z. This will require changes to RHG_graph and various SurfaceCode methods, as well as the check_correction function in decoder.py.

### PR DESCRIPTION
**Pull request by sduquemesa**
## Context for changes

- **New features**:
    * A new convention for boundaries have been introduced: the choice `"all_periodic"` refers to a code with boundaries periodic in all directions whereas codes with `"periodic"` have periodic boundaries in the _x_ and _y_ axis but not the _z_ axis.

## Example usage and tests

A surface code with all-periodic boundaries
```python
    from flamingpy.examples.surface_code import illustrate_surface_code

    params = {
        "d": 2, 
        "boundaries": "all_periodic",     # Boundaries
        "err": "primal",
        "polarity": None,
        "show": True,
    }

    illustrate_surface_code(**params)
```

A surface code with periodic boundaries in `x` and `y` but not `z`
```python
    from flamingpy.examples.surface_code import illustrate_surface_code

    params = {
        "d": 2, 
        "boundaries": "periodic",     # Boundaries
        "err": "primal",
        "polarity": None,
        "show": True,
    }

    illustrate_surface_code(**params)
```

## Performance results justifying changes

- [x] Not Applicable

## Workflow actions and tests

- [x] Not Applicable

## Expected benefits and drawbacks

**Expected benefits:**
- A more natural convention for periodic boundaries is put in place, giving user further option to configure their EC codes.

**Possible drawbacks:**
- None

## Related Github issues

Closes #52 

## Checklist and integration statements

- [x] My Python and C++ codes follow this project's coding and commenting styles as indicated by existing files. Precisely, the changes conform to given `black`, `docformatter` and `pylint` configurations. 
- [x] I have performed a self-review of these changes, checked my code (including for [codefactor](https://www.codefactor.io/repository/github/xanaduai/flamingpy/branches) compliance), and corrected misspellings to the best of my capacity. I have synced this branch with others as required.
- [x] I have added context for corresponding changes in documentation and [`README.md`](README.md) as needed.
- [x] I have added new workflow CI tests for corresponding changes, ensuring codecoverage is 95% or better, and these pass locally for me.
- [x] I have updated [`CHANGELOG.md`](.github/CHANGELOG.md) following the template. I recognize that the developers may revisit [`CHANGELOG.md`](.github/CHANGELOG.md) and the versioning, and create a Special Release including my changes.

### CODE REVIEW

**Comment by soosub** [comment]
Hey @ilan-tz I was just going over this PR trying to understand it. Wouldn't it be useful to also allow for one periodic axis? That case seems to be (physically) more relevant to me.

**Comment by ilan-tz** [comment]
> Hey @ilan-tz I was just going over this PR trying to understand it. Wouldn't it be useful to also allow for one periodic axis? That case seems to be (physically) more relevant to me.

We could do this. The top priority is periodic in x & y and finite in z because it's a more natural boundary choice (equating to toric code evolved in time) that what we currently have. For full flexibility we could also include periodic in just one of x or y -- @sduquemesa do you think it would be reasonable to include in this PR?

As for what's physically relevant -- if your hardware has the capability of creating a code with periodic boundaries in one dimension (e.g. via long-range interaction) then it can probably do it in the other as well!

**Comment by sduquemesa** [comment]
Hi @ilan-tz and @soosub! Please go ahead and give a first round of review to this PR, I'd appreciate your thorough examination. If the logic currently used here is good, we can implement periodicity in just one of the _x_ or _y_ axes. Thanks!

**Comment by soosub** [comment]
I'll have a look today 👍🏼

File: flamingpy/codes/surface_code.py [hunk 1]
```diff
@@ -444,24 +461,33 @@ def identify_boundary(self):
...
+            if "periodic" in self.bound_str:
+                ec_bound_points = []
             else:
-                dims = self.dims
-                syndrome_coords = getattr(self, ec + "_syndrome_coords")
                 bound_ind = np.where(self.boundaries == ec)[0][0]
-                plane_dict = {0: "x", 1: "y", 2: "z"}
+                ec_bound_points = self._get_ec_bounds(ec, bound_ind)
+
+            print(ec_bound_points)
```
[the remark this answers is not in the document]
Comment (soosub) [reply flamingpy/codes/surface_code.py:473]: We could provide an option for the user to print this while generating the SurfaceCode object **or** we could implement the boundary function as an `@property` method. I think the latter might be even more useful. We could do that (or something similar) for `identify_boundary`, `generate_stabilizer_coords`,  `six_body_dual_stabilizers` and probably also `periodic_boundary_stabilizers`.

**Review by soosub — COMMENTED** [review]
Here are some comments already. I tested the code by drawing various configurations and it seems to work at first glance. I'll have a more thorough look to see if the `ec_bound_points` are correct and if the tests make sense.

**Comment by soosub** [comment]
Just a note that the combination of `ec='primal'` and `boundaries='dual'` and the converse (`ec='dual'` and `boundaries='primal'`) result in a `KeyError`. I discussed this with Ilan and it should not work (so that's good 😃), but maybe it would be good to throw a more interpretable warning. 

Here is the whole Traceback:
```
KeyError                                  Traceback (most recent call last)
Input In [18], in <cell line: 2>()
      1 d, ec, boundary = (2, None, 'primal')
----> 2 c = SurfaceCode(d, ec=ec, boundaries=boundary)

File ~\flamingpy\flamingpy\codes\surface_code.py:315, in SurfaceCode.__init__(self, distance, ec, boundaries, polarity, backend)
    312 self.graph.index_generator()
    313 # The following line defines the stabilizer, syndrome coordinate,
    314 # and syndrome index attributes.
--> 315 self.identify_stabilizers()
    316 # The following line defines the boundary points attribute.
    317 self.identify_boundary()

File ~\flamingpy\flamingpy\codes\surface_code.py:357, in SurfaceCode.identify_stabilizers(self)
    354 all_six_bodies = self.generate_stabilizer_coords()
    356 # set {ec}_syndrome_coords, {ec}_syndrome_inds, and {ec}_stabilizers
--> 357 self.set_ec_stabilizers_and_syndrome(all_six_bodies)
    359 # set `all_syndrome_inds` and `all_syndrome_coords`
    360 for att in ["_syndrome_inds", "_syndrome_coords"]:

File ~\flamingpy\flamingpy\codes\surface_code.py:381, in SurfaceCode.set_ec_stabilizers_and_syndrome(self, all_six_bodies)
    379 all_cubes = []
    380 syndrome_coords = []
--> 381 for stabe in all_six_bodies[ec]:
    382     actual_stabe = list(set(stabe).intersection(rhg_lattice))
    383     # Dealing with stabilizers at periodic boundaries

KeyError: None
```

File: flamingpy/codes/surface_code.py [hunk 2]
```diff
@@ -88,18 +88,27 @@ def str_to_bound(bound_name):
 
         'open_primal': [primal, dual, dual]
         'open_dual': [primal, dual, primal]
+        'primal': [primal, primal, primal]
+        'dual': [dual, dual, dual]
+        'all_periodic': [periodic, periodic, periodic]
+        'periodic_primal': [periodic, periodic, primal]
+        'periodic_dual': [periodic, periodic, dual]
```
[the remark this answers is not in the document]
Comment (ilan-tz) [reply flamingpy/codes/surface_code.py:95]: In my opinion, the utility of string options is to give the user a convenient way to input common and meaningful choices of boundaries. Otherwise, they can input a list directly. Here, I would use the convention:

- If a single word, assume this choice of boundary in all three directions (i.e. replace `all_periodic` with `periodic`).
- If two words, assume common choice of boundaries for x, y (first word) and then boundary in z (second word). (i.e. everything else can stay).

For all other choices, a list suffices!

File: flamingpy/codes/surface_code.py [hunk 3]
```diff
@@ -285,6 +300,8 @@ def __init__(
 
         if boundaries == "open":
             self.bound_str = "open_primal" if ec in ("primal", "both") else "open_dual"
```
Comment (ilan-tz) [inline flamingpy/codes/surface_code.py:302]:
```suggestion
            self.bound_str = "open_dual" if ec in ("primal", "both") else "open_primal"
```
(Just to align with our option relabelling)

File: flamingpy/codes/surface_code.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilan-tz) [reply flamingpy/codes/surface_code.py:473]: These functions construct rather than simply return the boundary points, so let's avoid using `@property` here. The print statement should be removed.

File: flamingpy/codes/surface_code.py [hunk 4]
```diff
@@ -444,24 +461,33 @@ def identify_boundary(self):
...
-                dims = self.dims
-                syndrome_coords = getattr(self, ec + "_syndrome_coords")
                 bound_ind = np.where(self.boundaries == ec)[0][0]
-                plane_dict = {0: "x", 1: "y", 2: "z"}
+                ec_bound_points = self._get_ec_bounds(ec, bound_ind)
+
+            print(ec_bound_points)
+            setattr(self, ec + "_bound_points", ec_bound_points)
+
+    def _get_ec_bounds(self, ec, bound_ind):
```
Comment (ilan-tz) [inline flamingpy/codes/surface_code.py:476]: Could we have a short docstring here?

File: tests/decoders/test_decoder.py [hunk 5]
```diff
@@ -35,7 +35,7 @@
 
 
 code_params = it.product(
-    [2, 3, 4], ["primal", "dual"], ["open", "periodic"], [1, 0.1, 0.01], [0, 0.5, 1]
+    [2, 3, 4], ["primal", "dual"], ["open", "all_periodic"], [1, 0.1, 0.01], [0, 0.5, 1]
```
Comment (ilan-tz) [inline tests/decoders/test_decoder.py:38]:
```suggestion
    [2, 3, 4], ["primal", "dual"], ["open", "toric", "periodic"], [1, 0.1, 0.01], [0, 0.5, 1]
```

File: tests/decoders/test_decoder.py [hunk 6]
```diff
@@ -182,7 +182,7 @@ def test_correction_check(self, enc_state):
         failure_events = []
         for i, ec in enumerate(enc_state[0].ec):
             surface_dict = surface_dicts[i]
-            if boundaries == "periodic":
+            if boundaries == "all_periodic":
```
Comment (ilan-tz) [inline tests/decoders/test_decoder.py:185]:
```suggestion
            if boundaries == "periodic":
```

File: tests/decoders/test_uf_decoder.py [hunk 7]
```diff
@@ -62,7 +62,7 @@ def enc_state(request):
 
 
 code_params = it.product(
-    [2, 3, 4], ["primal", "dual"], ["open", "periodic"], [1, 0.1, 0.01]
+    [2, 3, 4], ["primal", "dual"], ["open", "all_periodic"], [1, 0.1, 0.01]
```
Comment (ilan-tz) [inline tests/decoders/test_uf_decoder.py:65]:
```suggestion
    [2, 3, 4], ["primal", "dual"], ["open", "toric", "periodic"], [1, 0.1, 0.01]
```

File: tests/qubit_codes/test_stabilizer_graph.py [hunk 8]
```diff
@@ -32,7 +32,7 @@
 
 code_params = it.product(
     [2, 3, 4],
-    ["open", "periodic"],
+    ["open", "all_periodic"],
```
Comment (ilan-tz) [inline tests/qubit_codes/test_stabilizer_graph.py:35]:
```suggestion
    ["open", "toric", "periodic"],
```

File: tests/qubit_codes/test_stabilizer_graph.py [hunk 9]
```diff
@@ -149,7 +149,7 @@ def test_shortest_paths_have_same_weight(enc_state):
 
 
 code_params2 = it.product(
-    [3, 4], ["open", "periodic"], ["primal", "dual"], [1, 0.9], [0, 0.5, 1], stab_graph_backend
+    [3, 4], ["open", "all_periodic"], ["primal", "dual"], [1, 0.9], [0, 0.5, 1], stab_graph_backend
```
Comment (ilan-tz) [inline tests/qubit_codes/test_stabilizer_graph.py:152]:
```suggestion
    [3, 4], ["open", "toric", "periodic"], ["primal", "dual"], [1, 0.9], [0, 0.5, 1], stab_graph_backend
```

File: tests/qubit_codes/test_surface_code.py [hunk 10]
```diff
@@ -70,6 +71,8 @@ def RHG_graph_old(dims, boundaries="finite", macronodes=False, polarity=False):
     # Dealing with boundaries
     if boundaries == "finite":
         boundaries = ["primal", "dual", "dual"]
+    elif boundaries == "all_periodic":
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:74]:
```suggestion
    elif boundaries == "periodic":
```

File: tests/qubit_codes/test_surface_code.py [hunk 11]
```diff
@@ -248,17 +251,17 @@ def test_boundary_combinations(self, d):
             # The following test requires a modification to RHG_graph_old,
             # since the new method creates a lattice with a different shape.
             # assert not set(RHG_lattice.edges) - set(RHG_graph_old(d, boundaries).edges)
-        for boundaries in ["primal", "dual", "periodic", "open_primal", "open_dual"]:
+        for boundaries in ["primal", "dual", "all_periodic", "open_primal", "open_dual"]:
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:254]:
```suggestion
        for boundaries in ["primal", "dual", "periodic", "open_primal", "open_dual"]:
```

File: tests/qubit_codes/test_surface_code.py [hunk 12]
```diff
@@ -248,17 +251,17 @@ def test_boundary_combinations(self, d):
...
         # assert not set(RHG_lattice_finite.edges) - set(RHG_graph_old(d, "finite").edges)
 
-    def test_periodic_boundaries(self, d):
-        """Test whether periodic boundary conditions produce a lattice with the
-        expected size."""
-        RHG_lattice = RHG_graph(d, "periodic")
+    def test_all_periodic_boundaries(self, d):
+        """Test whether periodic boundary conditions on every direction produce
+        a lattice with the expected size."""
+        RHG_lattice = RHG_graph(d, "all_periodic")
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:264]:
```suggestion
        RHG_lattice = RHG_graph(d, "periodic")
```

File: tests/qubit_codes/test_surface_code.py [hunk 13]
```diff
@@ -270,7 +273,36 @@ def test_periodic_boundaries(self, d):
         for point in all_boundaries:
             assert len(RHG_lattice[point]) == 4
 
-        assert not set(RHG_lattice.edges) - set(RHG_graph(d, "periodic").edges)
+        assert not set(RHG_lattice.edges) - set(RHG_graph(d, "all_periodic").edges)
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:276]:
```suggestion
        assert not set(RHG_lattice.edges) - set(RHG_graph(d, "periodic").edges)
```

File: tests/qubit_codes/test_surface_code.py [hunk 14]
```diff
@@ -310,7 +342,7 @@ def test_polarity(self, d, boundaries):
     def test_macronodes(self, d):
         """Check that the macronode lattice was generated as expected."""
         # Tests for periodic boundaries.
-        for boundaries in ["periodic"]:
+        for boundaries in ["all_periodic"]:
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:345]:
```suggestion
        for boundaries in ["periodic"]:
```

File: tests/qubit_codes/test_surface_code.py [hunk 15]
```diff
@@ -339,7 +371,7 @@ def test_macronodes(self, d):
             assert len(macronode_lattice) == 4 * len(RHG_reduced)
 
 
-code_params = it.product(range(2, 5), ["primal", "dual"], ["open", "periodic"])
+code_params = it.product(range(2, 5), ["primal", "dual"], ["open", "all_periodic"])
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:374]:
```suggestion
code_params = it.product(range(2, 5), ["primal", "dual"], ["open", "toric", "periodic"])
```

File: tests/qubit_codes/test_surface_code.py [hunk 16]
```diff
@@ -384,15 +416,15 @@ def test_stabilizers(self, surface_code):
                     assert len(cubes) == (d - 1) ** 2 * d
                 else:
                     assert len(cubes) == d**2 * (d - 1)
-            elif surface_code.bound_str == "periodic":
+            elif surface_code.bound_str == "all_periodic":
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:419]:
```suggestion
            elif surface_code.bound_str == "periodic":
```

File: tests/qubit_codes/test_surface_code.py [hunk 17]
```diff
@@ -384,15 +416,15 @@ def test_stabilizers(self, surface_code):
...
                 assert len(cubes) == d**3
             # Check that each stabilizer has 6 corresponing physical
             # vertices, even if it's an n-body stabilizer with n < 6.
             for cube in cubes:
                 assert len(cube.physical) == 6
                 # For periodic boundaries, check that there are
                 # only 6-body stabilizers.
-                if surface_code.bound_str == "periodic":
+                if surface_code.bound_str == "all_periodic":
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:427]:
```suggestion
                if surface_code.bound_str == "periodic":
```

File: tests/qubit_codes/test_surface_code.py [hunk 18]
```diff
@@ -413,7 +445,7 @@ def test_syndrome(self, surface_code):
                     assert len(syndrome_coords) == (d - 1) ** 3 + 2 * (d - 1) * d**2
                 else:
                     assert len(syndrome_coords) == d**3 + 2 * d * (d - 1) ** 2
-            elif surface_code.bound_str == "periodic":
+            elif surface_code.bound_str == "all_periodic":
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:448]:
```suggestion
            elif surface_code.bound_str == "periodic":
```

File: tests/qubit_codes/test_surface_code.py [hunk 19]
```diff
@@ -426,7 +458,7 @@ def test_boundary(self, surface_code):
                     assert len(bound_points) == 2 * d * (d - 1)
                 else:
                     assert len(bound_points) == 2 * d**2
-            elif surface_code.bound_str == "periodic":
+            elif surface_code.bound_str == "all_periodic":
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:461]:
```suggestion
            elif surface_code.bound_str in ("toric", "periodic"):
```

File: tests/qubit_codes/test_surface_code.py [hunk 20]
```diff
@@ -270,7 +273,36 @@ def test_periodic_boundaries(self, d):
         for point in all_boundaries:
             assert len(RHG_lattice[point]) == 4
 
-        assert not set(RHG_lattice.edges) - set(RHG_graph(d, "periodic").edges)
+        assert not set(RHG_lattice.edges) - set(RHG_graph(d, "all_periodic").edges)
+
+    @pytest.mark.parametrize("boundary_type", ["periodic_primal", "periodic_dual"])
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:278]: I would also test the "periodic" (all periodic) option here.

File: tests/qubit_codes/test_surface_code.py [hunk 21]
```diff
@@ -270,7 +273,36 @@ def test_periodic_boundaries(self, d):
...
+    def test_periodic_boundaries(self, d, boundary_type):
+        """Test whether periodic boundary conditions on x and y direction
+        produce a lattice with the expected neighbors."""
+        RHG_lattice = RHG_graph(d, boundary_type)
+        # assert len(RHG_lattice) == 6 * (d**3)
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:283]: Can this be brought back for the "periodic" setting check?

File: tests/qubit_codes/test_surface_code.py [hunk 22]
```diff
@@ -23,6 +23,7 @@
 import numpy as np
 from numpy.random import default_rng as rng
 import pytest
+from sympy import intersection
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:26]: What is the purpose in using sympy's intersection?

File: tests/qubit_codes/test_surface_code.py [hunk 23]
```diff
@@ -270,7 +273,36 @@ def test_periodic_boundaries(self, d):
         for point in all_boundaries:
             assert len(RHG_lattice[point]) == 4
 
-        assert not set(RHG_lattice.edges) - set(RHG_graph(d, "periodic").edges)
+        assert not set(RHG_lattice.edges) - set(RHG_graph(d, "all_periodic").edges)
+
+    @pytest.mark.parametrize("boundary_type", ["periodic_primal", "periodic_dual"])
+    def test_periodic_boundaries(self, d, boundary_type):
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:279]: I think this test should be modified. We should check whether there exists an edge between nodes on the low and high boundary at directions that are periodic. 

For example, if node "a" is on the low boundary, and node "b" is on the high boundary, and we expected them to be connected, we should ensure that the edge (a, b) exists in RHG_graph.edges.

**Review by ilan-tz — CHANGES_REQUESTED** [review]
Great work -- this is an important feature to include so that the user can enjoy a more traditional combination of boundary types.

There is a little bit of work to be done to relabel our options (if you think the suggestions make sense). The tests should also include the new boundary option to make sure that this choice propagates appropriately in the entire simulations pipeline (I've made suggestions of where to include the new option).

Lastly, a very important change needs to be made to `decoder.py`. Notice lines 215 to 218:

```
        if code.bound_str == "periodic":
            planes_to_check = ["x", "y", "z"]
        elif code.bound_str.startswith("open"):
            planes_to_check = ["x"] if ec == "primal" else ["y"]
```
We should include the following line:

```
    if code.bound_str == "toric":
            planes_to_check = ["x", "y"]
```
This means we will be checking that the toric code preserves the logical information for both qubits that it can encode.

(Side note: we are doing something a little unusual in the "periodic" case by checking if all three surfaces have the expected parity (i.e. error correction has succeed). We can keep this for historical reasons.)

File: flamingpy/codes/surface_code.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sduquemesa) [reply flamingpy/codes/surface_code.py:95]: Perfect! I'll apply this suggestions 🤓

File: flamingpy/codes/surface_code.py [hunk 24]
```diff
@@ -241,7 +255,8 @@ class SurfaceCode:
 
             'open': ['primal', 'dual', 'dual'] for 'primal' EC
                     ['primal', 'dual', 'primal'] for 'dual' EC
-            'periodic': 'periodic' in all three directions.
+            'all_periodic': 'periodic' in all three directions.
+            'periodic': 'periodic' in x and y but not z.
```
[the remark this answers is not in the document]
Comment (soosub) [reply flamingpy/codes/surface_code.py:259]: Yes! I like this suggestion 🍩

File: flamingpy/codes/surface_code.py [hunk 25]
```diff
@@ -88,18 +88,27 @@ def str_to_bound(bound_name):
...
-    elif bound_name == "open_dual":
-        boundaries = ["primal", "dual", "primal"]
-    elif bound_name in ("primal", "dual", "periodic"):
-        boundaries = [bound_name] * 3
-    elif not isinstance(bound_name, str):
-        print("Boundary type must be string.")
-        raise Exception
-    return np.array(boundaries)
+    if not isinstance(bound_name, str):
+        raise ValueError("Boundary type must be string.")
```
[the remark this answers is not in the document]
Comment (sduquemesa) [reply flamingpy/codes/surface_code.py:99]: yep

File: tests/qubit_codes/test_surface_code.py (hunk 22 above)
Comment (sduquemesa) [reply tests/qubit_codes/test_surface_code.py:26]: None — I blame the auto-import feature of my editor. Good catch! 👁️

File: flamingpy/codes/surface_code.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (sduquemesa) [reply flamingpy/codes/surface_code.py:257]: taking the donut for the title of this PR

File: tests/qubit_codes/test_surface_code.py [hunk 26]
```diff
@@ -272,6 +274,29 @@ def test_periodic_boundaries(self, d):
...
+                    coords for coords, props in neighbors.items() if "periodic" in props
+                ]
+
+                # check there is an edge from neighbor in low boundary to high boundary
+                # for all periodic neighbors
+                assert all(
+                    RHG_lattice.has_edge(point, periodic_point)
+                    for periodic_point in periodic_neighbors
+                )
+
```
Comment (ilan-tz) [inline tests/qubit_codes/test_surface_code.py:299]: Let us actually rethink this test. A few things:
- We may as well also test the `periodic` setting.
- Currently the test is trivial: it checks whether `point` and `periodic_point` have an edge, but they must have an edge by definition because they are neighbours of each other.
- I think we can recreate the logic from lines 225-234 of surface_code.py : do the nodes on the low boundary and the high boundary that we expect to have an edge between them, along with the ``periodic`` tag, actually have it. (Alternatively, we can do a counting argument: are there the expected number of periodic edges on each boundary. This will require a little bit of work to find the pattern).

File: .github/CHANGELOG.md [hunk 27]
```diff
@@ -2,6 +2,8 @@
 
 ### New features since the last release
 * Add functions to create different graph states (star and complete graphs, ring graphs, linear clusters, and Bell pairs) in a new module, `utils.graph_states`. [#68](https://github.com/XanaduAI/flamingpy/pull/68). (backward compatible)
+* A new convention for boundaries have been introduced: the choice `"all_periodic"` refers to a code with boundaries periodic in all directions whereas
+  codes with `"periodic"` have periodic boundaries in the _x_ and _y_ axis but not the _z_ axis. [#62](https://github.com/XanaduAI/flamingpy/pull/62). (backward compatible)
```
Comment (ilan-tz) [inline .github/CHANGELOG.md:6]: This part needs updating in light of latest changes.

**Review by ilan-tz — CHANGES_REQUESTED** [review]
Excellent updates since last review! Just a couple of things remaining:

- The test should be modified, per my comment. In addition, `test_correction_check` should be be modified in `test_decoder.py` (same line added as in 217 and 218 of `decoder.py`). The good news is the sanity checks seem to pass so I think the changes are all ok!
- The comments should be changed to add the new "toric" option in:
  - run_error_correction.py tutorial
  - surface_code.py example
    - In this example also: there is an `if` block that can be modified to include "toric". Note this could be good inspiration for the new test as well, since it's a visual check for additional edges. 
  - decoding.py benchmark

File: .github/CHANGELOG.md (hunk 27 above)
Comment (soosub) [inline .github/CHANGELOG.md:6]:
```suggestion
* A new convention for boundaries have been introduced: the choice `"periodic"` refers to a code with boundaries periodic in all (x, y and z) directions whereas
  codes with `"toric"` have periodic boundaries in the _x_ and _y_ axis but not the _z_ axis. [#62](https://github.com/XanaduAI/flamingpy/pull/62). (backward compatible)
```

File: flamingpy/codes/surface_code.py [hunk 28]
```diff
@@ -239,9 +253,16 @@ class SurfaceCode:
...
         boundaries (str): the boundary conditions. The options are:
 
-            'open': ['primal', 'dual', 'dual'] for 'primal' EC
-                    ['primal', 'dual', 'primal'] for 'dual' EC
+            'open': ['primal', 'dual', 'dual'] for 'primal' EC,
+                    ['primal', 'dual', 'primal'] for 'dual' EC.
             'periodic': 'periodic' in all three directions.
+            'toric':  ['periodic', 'periodic', 'dual'] for 'primal' EC,
+                      ['periodic', 'periodic', 'primal'] for 'dual' EC.
+
```
Comment (soosub) [inline flamingpy/codes/surface_code.py:261]: This looks a bit weird in the rendered Sphinx documentation. It might be worth fixing it.

File: flamingpy/codes/surface_code.py [hunk 29]
```diff
@@ -298,9 +304,14 @@ def __init__(
         self.ec = [ec]
         # self.ec = ["primal", "dual"] if ec == "both" else [ec]
 
+        if {ec, boundaries} == {"primal", "dual"}:
```
[the remark this answers is not in the document]
Comment (sduquemesa) [reply flamingpy/codes/surface_code.py:307]: What I am trying to catch here is the condition [mentioned by @soosub](https://github.com/XanaduAI/flamingpy/pull/62#issuecomment-1151480286) on which an uninformative error is raised

> Just a note that the combination of ec='primal' and boundaries='dual' and the converse (ec='dual' and boundaries='primal') result in a KeyError. I discussed this with Ilan and it should not work (so that's good 😃), but maybe it would be good to throw a more interpretable warning.

This suggestion does improve the sanity check. Thanks!

File: flamingpy/codes/surface_code.py (hunk 28 above)
Comment (sduquemesa) [reply flamingpy/codes/surface_code.py]: This is how it looks now: 👀

image

File: tests/qubit_codes/test_surface_code.py (hunk 26 above)
Comment (sduquemesa) [reply tests/qubit_codes/test_surface_code.py:299]: > I think we can recreate the logic from lines 225-234 of surface_code.py : do the nodes on the low boundary and the high boundary that we expect to have an edge between them, along with the periodic tag, actually have it. (Alternatively, we can do a counting argument: are there the expected number of periodic edges on each boundary. This will require a little bit of work to find the pattern).

Does it actually make sense to test using the same logic? I'd be more in favor of using an unrelated logic, i.e., _are there the expected number of periodic edges on each boundary_; however I will need some help to figure the pattern out.

Comment (sduquemesa) [reply tests/qubit_codes/test_surface_code.py]: Done! 💪

**Review by ilan-tz — APPROVED** [review]
Beautiful! I've added a small part to the test to check if the expected periodic edges have the "periodic" attribute set to True.
